stdlib/public/core/HashedCollections.swift.gyb

```diff
@@ -2187,6 +2187,24 @@ extension Dictionary {
       _variantBuffer: _variantBuffer.mapValues(transform))
   }
 
+  /// Returns a new dictionary containing the keys of this dictionary with the
+  /// values transformed by the given closure.
+  /// - Parameter transform: A closure that transforms a value. `transform`
+  ///   accepts each value of the dictionary as its parameter and returns a
+  ///   transformed value of the same or of a different type.
+  /// - Returns: A dictionary containing the keys and transformed values of
+  ///   this dictionary.
+  @_inlineable // FIXME(sil-serialize-all)
+  public func compactMapValues<T>(
+    _ transform: (Value) throws -> T?
+  ) rethrows -> Dictionary<Key, T> {
+    return try self.reduce(into: [Key: T](), { (result, x) in
+      if let value = try transform(x.value) {
+        result[x.key] = value
+      }
```

Comment:
I would've expected it to compose using `Dictionary.init(uniqueKeysWithValues:)`, or if nothing else using `Dictionary(minimumCapacity:)` as the argument to `reduce` as an optimization.

Reply:
This is due to an intentional tradeoff between space and time. Capacity reservation works best when we have a clear *lower* bound on how many elements we’re going to add. `self.count` is an upper bound to the number of elements in the resulting dictionary, but we don’t have a lower bound, and allocating that much space would often be too wasteful.